File: python/evm_state/history.py
"""Remove whole native history partitions after verified checkpoint publication.

Each source is bound to one checkpoint database. All of that database's retained
checkpoints for these accounts protect their continuation intervals. This cleans
already checkpointed history; initial-history compaction is a separate operation.
"""
import json
from pathlib import Path

from .checkpoint import _manifest, connect_like
from .control import control
from .cursor import load_progress
from .files import exclusive_lock
from .proof import VerificationError
from .source import verified_source

PARTITIONS = {"state_blocks": "toDate(_block_timestamp_)", "_blocks_": "toYYYYMM(timestamp)"}


def _plan(client, checkpoints, run, checked, snapshot_id, directory, keep_blocks):
    if isinstance(keep_blocks, bool) or not isinstance(keep_blocks, int) or keep_blocks < 1:
        raise ValueError("keep_blocks must be at least one")
    if run["identity"]["checkpoint_database"] != checkpoints.database:
        raise VerificationError("native history belongs to another checkpoint database")
    ready = _manifest(checkpoints, snapshot_id)
    covered = [source for source in ready["sources"] if source.get("run_id") == run["run_id"]]
    provenance = ["database_uuid", "package_sha256", "schema_hash", "module_hash", "accounts"]
    if len(covered) != 1 or any(covered[0].get(key) != checked.get(key) for key in provenance):
        raise VerificationError("checkpoint does not cover this exact native source")
    selected = set(run["identity"]["accounts"])
    if not selected.issubset(ready["accounts"]):
        raise VerificationError("checkpoint does not cover all source accounts")
    progress = load_progress(client, run, directory)
    tip = progress["position"]["block"]
    if (Path(directory) / "cursor.txt").read_text().strip() != progress["cursor"]:
        raise VerificationError("native cursor differs from durable progress; recover or finish the run before cleanup")
    if tip["number"] < ready["header"]["number"]:
        raise VerificationError("native progress has not reached the checkpoint")
    # Preserve updates following every retained checkpoint, including pins and
    # checkpoints imported from another source. Publication/rotation is excluded
    # while this list and the cleanup plan are in use.
    floor = ready["header"]["number"]
    protected = []
    for row in checkpoints.rows("SELECT snapshot_id,manifest FROM checkpoints FINAL"):
        record = json.loads(row["manifest"])
        if record.get("status") != "ready" or record.get("snapshot_id") != row["snapshot_id"]:
            raise VerificationError("invalid retained checkpoint prevents native history cleanup")
        if selected.intersection(record["accounts"]):
            floor = min(floor, int(record["header"]["number"]))
            protected.append(record["snapshot_id"])
    remove_before = min(floor + 1, tip["number"] - keep_blocks + 1)
    schemas = {r["name"]: r["partition_key"] for r in client.rows(
        "SELECT name,partition_key FROM system.tables WHERE database={db:String} "
        "AND name IN ('state_blocks','_blocks_')", {"db": client.database})}
    if schemas != PARTITIONS:
        raise VerificationError("native history schema does not match the qualified partition layout")
    tables = {}
    for table in PARTITIONS:
        tables[table] = list(client.rows(f"SELECT _partition_id AS id,min(number) AS first,max(number) AS last "
            f"FROM {table} GROUP BY _partition_id HAVING last < {{before:UInt64}} ORDER BY first",
            {"before": max(0, remove_before)}))
    return {"source": checked, "checkpoint": snapshot_id, "protected_checkpoints": sorted(protected),
            "remove_before": max(0, remove_before), "keep_blocks": keep_blocks,
            "durable_block": tip, "partitions": tables}


def cleanup(client, directory, snapshot_id, keep_blocks=10000, apply=False):
    directory = Path(directory).resolve()
    run = json.loads((directory / "run.json").read_text())
    identity = run["identity"]
    if identity.get("format_version") != 3:
        raise VerificationError("native cleanup requires destination-bound run identity format 3")
    checkpoints = connect_like(client, identity["checkpoint_database"])
    source = {key: identity[key] for key in ["database", "accounts", "start_block", "module_hash", "final_blocks_only"]}
    # Same order as publication. Both writer locks fail instead of deadlocking
    # with a reader/build in another process; raw SQL bypasses this protocol.
    with control(checkpoints).publisher(), exclusive_lock(directory / "run.lock"):
        with verified_source(client, source, checkpoints) as checked:
            if checked["state_directory"] != str(directory) or checked["run_id"] != run["run_id"]:
                raise VerificationError("cleanup directory does not own this native source")
        with exclusive_lock(directory / "source_readers.lock"):
            result = _plan(client, checkpoints, run, checked, snapshot_id, directory, keep_blocks)
            result["bytes_before"] = client.disk_usage()
            if apply:
                for table, partitions in result["partitions"].items():
                    for partition in partitions:
                        client.execute(f"ALTER TABLE {table} DROP PARTITION ID {{id:String}}", {"id": partition["id"]})
                # The cursor's block and marker must still be present after all
                # drops. Interrupted cleanup can recompute the plan and resume.
                load_progress(client, run, directory)
            result.update(applied=apply, bytes_after=client.disk_usage(),
                space_reclamation="inactive parts may remain until ClickHouse background cleanup; byte counts include them")
            return result

File: crates/evm-state/src/history.rs
//! Whole native-history partition cleanup with retained continuation protection.
use crate::{
    ch::{params, uint, ClickHouse},
    checkpoint::{canonical_accounts, manifest},
    control::{Control, PublisherLock},
    cursor,
    files::{file_lock, resolve, Lock},
    proof::string,
    source::verified_source,
};
use anyhow::{ensure, Context, Result};
use serde_json::{json, Value};
use std::{
    collections::BTreeSet,
    fs,
    path::{Path, PathBuf},
};

const PARTITIONS: [(&str, &str); 2] = [
    ("state_blocks", "toDate(_block_timestamp_)"),
    ("_blocks_", "toYYYYMM(timestamp)"),
];

pub fn partitions_before(client: &ClickHouse, before: u64) -> Result<Value> {
    let mut schemas = serde_json::Map::new();
    for row in client.rows("SELECT name,partition_key FROM system.tables WHERE database={db:String} AND name IN ('state_blocks','_blocks_')", &params(json!({"db":client.database}))?)? {
        let row = row?;
        ensure!(schemas.insert(string(&row,"name")?.into(), row["partition_key"].clone()).is_none(), "duplicate native history table");
    }
    ensure!(
        Value::Object(schemas)
            == json!(PARTITIONS
                .into_iter()
                .collect::<std::collections::BTreeMap<_, _>>()),
        "native history schema does not match the qualified partition layout"
    );
    let mut result = json!({});
    for (table, _) in PARTITIONS {
        result[table] = json!(client.rows(&format!("SELECT _partition_id AS id,min(number) AS first,max(number) AS last FROM {table} GROUP BY _partition_id HAVING last < {{before:UInt64}} ORDER BY first"), &params(json!({"before":before}))?)?.collect::<Result<Vec<_>>>()?);
    }
    Ok(result)
}

pub(crate) fn drop_partitions(client: &ClickHouse, partitions: &Value) -> Result<()> {
    for (table, _) in PARTITIONS {
        for partition in partitions[table]
            .as_array()
            .context("invalid partition plan")?
        {
            client.execute(
                &format!("ALTER TABLE {table} DROP PARTITION ID {{id:String}}"),
                &params(json!({"id":string(partition,"id")?}))?,
            )?;
        }
    }
    Ok(())
}

pub(crate) fn declaration(identity: &Value) -> Result<Value> {
    let mut source = json!({});
    for key in [
        "database",
        "accounts",
        "start_block",
        "module_hash",
        "final_blocks_only",
    ] {
        source[key] = identity
            .get(key)
            .context("incomplete native source identity")?
            .clone();
    }
    Ok(source)
}

/// Lock order matches checkpoint publication and excludes both native writers and
/// checkpoint readers of this source. The initial shared lock is dropped before
/// the exclusive reader lock is acquired; the run lock prevents identity changes.
pub(crate) struct SourceWriter {
    pub directory: PathBuf,
    pub run: Value,
    pub source: Value,
    pub checkpoints: ClickHouse,
    pub control_path: PathBuf,
    _publisher: PublisherLock,
    _run: Lock,
    _readers: Lock,
}
impl SourceWriter {
    pub fn open(client: &ClickHouse, directory: &Path, setup: bool) -> Result<Self> {
        let directory = resolve(directory)?;
        let run: Value = serde_json::from_slice(&fs::read(directory.join("run.json"))?)?;
        ensure!(
            run["identity"]["format_version"] == 3,
            "native cleanup requires destination-bound run identity format 3"
        );
        let checkpoints = client.with_database(string(&run["identity"], "checkpoint_database")?)?;
        if setup {
            crate::checkpoint::setup(&checkpoints)?;
        }
        let control = Control::open(&checkpoints)?;
        let publisher = control.publisher()?;
        let run_lock = file_lock(&directory.join("run.lock"), true, false)?;
        let checked = verified_source(
            client,
            &declaration(&run["identity"])?,
            Some(&checkpoints),
            None,
        )?;
        ensure!(
            checked.source["state_directory"]
                == directory.to_str().context("invalid source directory")?
                && checked.source["run_id"] == run["run_id"],
            "cleanup directory does not own this native source"
        );
        let source = checked.source.clone();
        drop(checked);
        let readers = file_lock(&directory.join("source_readers.lock"), true, false)?;
        Ok(Self {
            directory,
            run,
            source,
            checkpoints,
            control_path: control.path,
            _publisher: publisher,
            _run: run_lock,
            _readers: readers,
        })
    }
    pub fn progress(&self, client: &ClickHouse) -> Result<Value> {
        let progress = cursor::load_progress(client, &self.run, &self.directory)?;
        ensure!(
            fs::read_to_string(self.directory.join("cursor.txt"))?.trim()
                == string(&progress, "cursor")?,
            "native cursor differs from durable progress; recover or finish the run before cleanup"
        );
        Ok(progress)
    }
}

fn plan(
    client: &ClickHouse,
    owner: &SourceWriter,
    snapshot_id: &str,
    keep_blocks: u64,
) -> Result<Value> {
    ensure!(keep_blocks > 0, "keep_blocks must be at least one");
    let ready = manifest(&owner.checkpoints, snapshot_id)?;
    let covered = ready["sources"]
        .as_array()
        .context("invalid checkpoint sources")?
        .iter()
        .filter(|s| s.get("run_id") == owner.run.get("run_id"))
        .collect::<Vec<_>>();
    ensure!(
        covered.len() == 1
            && [
                "database_uuid",
                "package_sha256",
                "schema_hash",
                "module_hash",
                "accounts"
            ]
            .iter()
            .all(|key| covered[0].get(key) == owner.source.get(key)),
        "checkpoint does not cover this exact native source"
    );
    let selected = canonical_accounts(&owner.run["identity"]["accounts"])?
        .into_iter()
        .collect::<BTreeSet<_>>();
    ensure!(
        selected.is_subset(
            &canonical_accounts(&ready["accounts"])?
                .into_iter()
                .collect()
        ),
        "checkpoint does not cover all source accounts"
    );
    let progress = owner.progress(client)?;
    let tip = uint(&progress["position"]["block"]["number"])?;
    let mut floor = uint(&ready["header"]["number"])?;
    ensure!(
        tip >= floor,
        "native progress has not reached the checkpoint"
    );
    let mut protected = Vec::new();
    for row in owner.checkpoints.rows(
        "SELECT snapshot_id,manifest FROM checkpoints FINAL",
        &Default::default(),
    )? {
        let row = row?;
        let record: Value = serde_json::from_str(string(&row, "manifest")?)?;
        ensure!(
            record["status"] == "ready" && record["snapshot_id"] == row["snapshot_id"],
            "invalid retained checkpoint prevents native history cleanup"
        );
        let accounts = canonical_accounts(&record["accounts"])?;
        if accounts.iter().any(|a| selected.contains(a)) {
            floor = floor.min(uint(&record["header"]["number"])?);
            protected.push(string(&record, "snapshot_id")?.to_owned());
        }
    }
    protected.sort();
    let before = floor
        .saturating_add(1)
        .min(tip.saturating_sub(keep_blocks - 1));
    Ok(
        json!({"source":owner.source,"checkpoint":snapshot_id,"protected_checkpoints":protected,"remove_before":before,"keep_blocks":keep_blocks,"durable_block":progress["position"]["block"],"partitions":partitions_before(client,before)?}),
    )
}

pub fn cleanup(
    client: &ClickHouse,
    directory: &Path,
    snapshot_id: &str,
    keep_blocks: u64,
    apply: bool,
) -> Result<Value> {
    ensure!(keep_blocks > 0, "keep_blocks must be at least one");
    let owner = SourceWriter::open(client, directory, false)?;
    let mut result = plan(client, &owner, snapshot_id, keep_blocks)?;
    result["bytes_before"] = json!(client.disk_usage()?);
    if apply {
        drop_partitions(client, &result["partitions"])?;
        owner.progress(client)?;
    }
    result["applied"] = json!(apply);
    result["bytes_after"] = json!(client.disk_usage()?);
    result["space_reclamation"] = json!(
        "inactive parts may remain until ClickHouse background cleanup; byte counts include them"
    );
    Ok(result)
}
